executor/src/constant_evaluator/mod.rs

```diff
@@ -555,19 +568,21 @@ mod test {
         let src = r#"
             constant %N = 4;
             namespace F(%N);
-            let x: col = |i| y(i) + 1;
-            let y: col = |i| i + 20;
+            let x = |i| y(i) + 1;
+            let y = |i| i + 20;
+            let X: col = x;
```

Comment:
so this is building a col from a generic fn that becomes concrete on `fe`?

Reply:
No, symbols are only generic if they are explicitly generic, i.e. `let<T> x: T -> T = ...`. If you do not specify a type, then a unique concrete type must be inferred from the way the symbol is used and from its definition. If the type is not unique (mostly meaning it is generic), the type checker fails.

What happens here is that the type of the value of a column can either be `int -> fe` or `int -> int` ('int -> fe' is preferred, if that fails, the type checker tries `int -> int`). Since the type of `y` is at least `T -> T`, the type of `x` cannot be `int -> fe` and thus the types of both `x` and `y` are inferred to be `int -> int`.

A reference to a column is `expr`. I think these simple rules allow a quite intuitive use without the need for implicit conversions everywhere. In the end, there is no real implicit conversion happening anyway.

Reply:
I don't think declaring a variable twice should be allowed. Or rather: we should decide if the language is supposed to remain declarative or not

Reply:
It is currently not allowed to declare two symbols of the same name in the same scope. The example above has `x` and `X`.